numpy/testing/_private/utils.py

```diff
@@ -244,34 +244,40 @@ def jiffies(_load_time=[]):
 
 
 def build_err_msg(arrays, err_msg, header='Items are not equal:',
-                  verbose=True, names=('ACTUAL', 'DESIRED'), precision=8):
+                  verbose=1, names=('ACTUAL', 'DESIRED'), precision=8):
     msg = ['\n' + header]
     if err_msg:
         if err_msg.find('\n') == -1 and len(err_msg) < 79-len(header):
             msg = [msg[0] + ' ' + err_msg]
         else:
             msg.append(err_msg)
     if verbose:
-        for i, a in enumerate(arrays):
-
-            if isinstance(a, ndarray):
-                # precision argument is only needed if the objects are ndarrays
-                r_func = partial(array_repr, precision=precision)
-            else:
-                r_func = repr
-
-            try:
-                r = r_func(a)
-            except Exception as exc:
-                r = '[repr failed for <{}>: {}]'.format(type(a).__name__, exc)
-            if r.count('\n') > 3:
-                r = '\n'.join(r.splitlines()[:3])
-                r += '...'
-            msg.append(' %s: %s' % (names[i], r))
+        from numpy.core import get_printoptions, printoptions, inf
+        options = get_printoptions()
+        if verbose > 1:
+            # Use get/update to avoid overwriting a custom formatter.
+            options.update(threshold=inf)
+        with printoptions(**options):
+            for i, a in enumerate(arrays):
+                if isinstance(a, ndarray):
+                    # Precision is only needed if the objects are ndarrays.
+                    r_func = partial(array_repr, precision=precision)
+                else:
+                    r_func = repr
+
+                try:
+                    r = r_func(a)
+                except Exception as exc:
+                    r = '[repr failed for <{}>: {}]'.format(type(a).__name__,
+                                                            exc)
+                if verbose < 2 and r.count('\n') > 3:
```

Comment:
The main problem with printing array seems to be that although the number of items is configurable in print options with threshold, the line numbers (number of lines that array takes) is not configurable.  It seems to be hard coded here with a value 3 and truncated if its greater than that. 

IMO the best way to address your use case is to add an option to set_printoptions to make the number of output lines printed for an array configurable, and the functions that depend on it should handle this option. This way we can get rid of the > 3 check here since the array repr would handle that.

Disclaimer, I am new to the numpy codebase and don't have full context on whether there are more reasons to not make this configurable and if there is a better way to implement your use case. 

Having said that, from what I have observed, such enhancements get more attention if there is an issue opened with the background and the proposed implementation and getting opinions of maintainers, before opening a PR.

Reply:
I have to agree that the number of lines with 3 here seems a bit arbitrary? We may be able to add a similar option to printoptions itself?

One thing I would like to know is whether it is not good enough to change your options globally in a setup of your pytest run, or maybe tweak the output here more generally.

Reply:
Hi, thanks for the comments!

I agree that 3 lines is completely arbitrary; the original code just had the check for `if r.count('\n') > 3`, so I just kept it, trying to modify as little as possible.

The reason I originally avoided making changes to `printoptions` directly was because this is testing-specific, and not a general property that `ndarray` always obeys on `repr`.  It's also perhaps not so nice to have `threshold` and `testing_max_lines` (or something) both in `printoptions`, since they're two different forms of truncation.

I inserted the call to `set_printoptions` above to fit in with the general idea of a higher level of verbosity.  I'm not trying to bend `numpy` to our use-case only, but since `testing.assert*` are testing functions, it would be strange to me as an end-user if I asked for "more verbose" output, and then the library truncated the output at a different point.

All that said, if you think that adding an extra line to `printoptions` is a better way to go, then I'll close this PR and resubmit.  As @anirudh2290 pointed out, it would have been better for me to open an issue for this first, but I guess we can use this PR to discuss now that I have opened it.

Would you be open to having the line-based truncation removed completely (i.e. take it out, not move it to `printoptions`), and just rely on the general `repr` method, or would it be better to add a `testing_max_lines` option?

Reply:
Ah, I did not quite realize that in a sense the printoptions part is an additional aside here, and the actual trunction you like is the 3 lines. Its tricky, IMO pytest output is already annoyingly verbose – at some point we need html output and folding to read it :) –, making it more verbose by removing the 3 line limit has its up and downsides, I do not have much of an opinion on it to be honest.

I am not convinced about the printoptions, and frankly, few probably pass the `verbose` kwarg, so if we can find a default that works for most users, that seems nicer.
Maybe another question would be if we can somehow make the message more informative for your use case without actually printing the full arrays.

Reply:
The `verbose` default used to be `True`, so I'd reckon that the number of people passing `False` is very very low.  I'd added an extra level on top so as not to disturb the default for the entire `numpy` testing suite.

The potential improvement I suppose would be to show items which violate the assertion, in context.  Off the top of my head, I can see two possible sticking points with this:
1. The assertion functions will have to pass in the indices of the values for which it has failed, but the `assert_*` functions are also intended to handle all input types (not just `ndarray`).
2. The concept of context is difficult to define a sensible default for that works for arbitrary dimensions.  It's easy enough with one or two, but much harder after that.  I suppose we could just output 3 flat columns - the indices of each failed point, and the values in each array (and ignore context).

I'm not sure that this method fits in without a larger modification to `numpy.testing` output.  I know lots of packages use these functions for their own assertions, but I'm not sure how much this is really supported/encouraged by `numpy`, and how much they're just meant to be for internal testing.  If it's more the latter, then I think a backwards-compatible modification to the `verbose` kwarg is the path of least resistance, if you're open to accepting this idea.  If this is the case, I'd understand if you didn't want to accept it.

If `numpy.testing` is really meant to be for external use as well, then it might be good to have a much more general approach to configuration, something like larger modifications to `printoptions`?

Reply:
One issue I have with the current proposed change to the `verbose` kwarg is that is "special-cases" the value `verbose=1` to be interpreted as a bool, so you get non-intuitive behavior:

verbose=1 -> truncate to **3** lines
verbose=n (where n > 1) -> truncate to **n** lines

With this approach, a user wouldn't be able to toggle the truncation to a single line, because `1` means `3`.

Rather than change the meaning of `verbose`, I think it would be more clear to introduce a new kwarg to testing to control the line truncation, e.g. `repr_max_lines=3`. The default behavior then remains the same, `verbose` doesn't conflate multiple meanings, and the line truncation becomes configurable.